### PR TITLE
Restore compatibility with models handling fewer tokens

### PR DESCRIPTION
**Pull request by willdumm**
This PR addresses #102. It still needs a unit test (started) and ideally at least a plan for how to apply the neutral model when we add additional tokens (there's some discussion in the issue about this).

### CODE REVIEW

File: netam/models.py [hunk 1]
```diff
@@ -597,11 +597,28 @@ def selection_factors_of_aa_str(self, aa_str: str) -> Tensor:
         mask = aa_mask_tensor_of(aa_str)
         mask = mask.to(self.device)
 
+        # Here we're ignoring sites containing tokens that have index greater
+        # than the embedding dimension.
+        if "embedding_dim" in self.hyperparameters:
+            consider_sites = aa_idxs < self.hyperparameters["embedding_dim"]
```
Comment (matsen) [inline netam/models.py:603]: I'm not sure how to read `consider_sites`. I think these are `meaningful_sites`? or `aa_sites`? 

I guess I'm a little confused about what this section of code is doing. It seems like we're making a big empty tensor, partially filling it, then trimming it back again.

File: netam/sequences.py [hunk 2]
```diff
@@ -17,9 +17,9 @@
 
 NT_STR_SORTED = "".join(BASES)
 BASES_AND_N_TO_INDEX = {base: idx for idx, base in enumerate(NT_STR_SORTED + "N")}
-# ambiguous must remain last. It is assumed elsewhere that the max index
-# denotes the ambiguous base
-AA_TOKEN_STR_SORTED = AA_STR_SORTED + RESERVED_TOKENS + "X"
+# Must add new tokens to the end of this string.
+AA_TOKEN_STR_SORTED = AA_STR_SORTED + "X" + RESERVED_TOKENS
```
Comment (matsen) [inline netam/sequences.py:21]: I suggest renaming

```suggestion
TOKEN_STR_SORTED = AA_STR_SORTED + "X" + RESERVED_TOKENS
```

File: tests/test_backward_compat.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import torch
+
+from netam.framework import load_crepe
+from netam.sequences import set_wt_to_nan
+
+
+# The outputs used for this test are produced by running
```
Comment (matsen) [inline tests/test_backward_compat.py:7]: This is great but perhaps give a little more context for when I forget what old means?

**Review by matsen — COMMENTED** [review]
Nice! A few comments / questions here.

File: netam/models.py (hunk 1 above)
Comment (willdumm) [reply netam/models.py:603]: I renamed to `model_valid_sites`.

This code is applying the model to an aa string, only considering the entries in that string that the model was trained to handle (marked by `model_valid_sites`). We construct a big empty `result` tensor of `nan`s, then feed the amino acid string to the model, stripping out sites containing tokens that aren't valid for the model, then put the model's outputs into the `result` tensor at the sites that contain tokens that we fed to the model.

For instance, if we have a heavy chain sequence `QQQQ^`, then the model will see `QQQQ`, but the output will have first dimension size 5, matching the input sequence length, and the last output will be `nan`s.

If we add a start token in the future, then we can still feed sequences containing it to our model. For instance, if we have `1QQQQ`, the old model not handling the new `1` token will see `QQQQ`, but the returned model output will contain `nan`s in the first site, and will have first dimension size 5, matching the input sequence length.

**Comment by willdumm** [comment]
The companion PR for this is now https://github.com/matsengrp/dnsm-experiments-1/pull/77

File: netam/models.py (hunk 1 above)
Comment (matsen) [reply netam/models.py]: Great!

Let's make a free function that encapsulates the test

```
if "embedding_dim" in self.hyperparameters:
```

with a nice name that we can look for when we strip this if out. I'm anticipating that we'll have a release in which we assume that all models have an `embedding_dim`.

(In other news, we could just go in and add an embedding dim into all the old models and assume they have them, right?)

**Review by matsen — COMMENTED** [review]
one little pending suggestion!

**Review by matsen — APPROVED** [review]

File: netam/models.py (hunk 1 above)
Comment (willdumm) [reply netam/models.py]: I addressed this by adding an `embedding_dim` to the Single model.
